### ISSUE TITLE
src/scanner.cc:126:13: error: use of undeclared identifier 'advance' advance(lexer);

### ISSUE THREAD

**Issue by yesudeep** [NTBBloodbath/doom-nvim#222]
When trying to use the develop branch on Mac OS:

```
[nvim-treesitter] [0/1] Compiling...
nvim-treesitter[norg]: Error during compilation
src/scanner.cc:106:39: error: expected expression
        return std::vector<TokenType>({ lhs, static_cast<TokenType>(rhs) });
                                      ^
src/scanner.cc:109:27: warning: rvalue references are a C++11 extension [-Wc++11-extensi
ons]
    std::vector<TokenType>&& operator|(std::vector<TokenType>&& lhs, TokenType rhs)
                          ^
src/scanner.cc:109:62: warning: rvalue references are a C++11 extension [-Wc++11-extensi
ons]
    std::vector<TokenType>&& operator|(std::vector<TokenType>&& lhs, TokenType rhs)
                                                             ^
src/scanner.cc:360:41: error: expected ';' at end of declaration list
    std::vector<size_t>& get_tag_stack() noexcept { return m_TagStack; }
                                        ^
                                        ;
src/scanner.cc:583:27: warning: default member initializer for non-static data member is
 a C++11 extension [-Wc++11-extensions]
    TokenType m_LastToken = NONE;
                          ^
src/scanner.cc:586:26: warning: default member initializer for non-static data member is
 a C++11 extension [-Wc++11-extensions]
    size_t m_ParsedChars = 0, m_IndentationLevel = 0;
                         ^
src/scanner.cc:586:50: warning: default member initializer for non-static data member is
 a C++11 extension [-Wc++11-extensions]
    size_t m_ParsedChars = 0, m_IndentationLevel = 0;
                                                 ^
src/scanner.cc:592:54: warning: default member initializer for non-static data member is
 a C++11 extension [-Wc++11-extensions]
    const std::array<int32_t, 8> s_DetachedModifiers = { '*', '-', '>', '|', '=', '~', '
$', '_' };
                                                     ^
src/scanner.cc:126:13: error: use of undeclared identifier 'advance'
            advance(lexer);
            ^
src/scanner.cc:133:13: error: use of undeclared identifier 'advance'
            advance(lexer);
            ^
src/scanner.cc:148:13: error: use of undeclared identifier 'advance'
            advance(lexer);
            ^
src/scanner.cc:158:17: error: use of undeclared identifier 'advance'
                advance(lexer);
                ^
src/scanner.cc:165:21: error: use of undeclared identifier 'advance'
                    advance(lexer);
                    ^
src/scanner.cc:177:21: error: use of undeclared identifier 'advance'
                    advance(lexer);
```

**Comment by connorgmeehan** [comment]
I'm pretty sure this is an issue with the neorg treesitter module, I have just disabled this module in `doom_modules.lua`

**Comment by NTBBloodbath** [comment]
Hi, yeah @connorgmeehan is right. It's more a nvim-treesitter issue with neorg parser atm because it's not officially supported yet by treesitter plugin.

You can temporarily disable it as connor said, look for a `neorg` entry in your `doom_modules.lua` and comment it.

**Comment by yesudeep** [comment]
Alright, makes sense. 

Aside: This can also be fixed by adding a `-std=c++11` or `-std=c++17` flag for the C++ compiler. I just don't know where to add it.

**Comment by connorgmeehan** [comment]
I've done some investigating into the issue, basically the treesitter parser doesn't compile with clang.  The workaround referenced [here](https://github.com/nvim-neorg/neorg/issues/74#issuecomment-906627223) is to install it with `gcc` instead.
1. Install gcc
    - `brew install gcc`
2. Get the gcc bin directory for 
    - `brew info gcc`
3. Backup default gcc (it may not exist but worth doing just incase
    - `mv /usr/local/bin/gcc /usr/local/bin/gcc_old`
4. Make a symlink to `/usr/local/bin`
    - `ln -s /usr/local/Cellar/gcc/{{version_number}}/bin/gcc-11 /usr/local/bin/gcc`
5. Restart neovim with `neorg` enabled in doom_modules, let it install neorg parser for you
6. Remove gcc
    - `rm /usr/local/bin/gcc && mv /usr/local/bin/gcc_old /usr/local/bin/gcc`
    - `brew uninstall gcc`

Let me know if this works for you! :)

**Comment by molleweide** [comment]
I am assuming this is related to M1 since it is not mentionned??

**Comment by molleweide** [comment]
This is only pertaining to M1, right?

**Comment by connorgmeehan** [comment]
I get the same error on my intel MBP and following these steps solved the issue for me.  Also investigating further, looks like tree-sitter parsers have better compatibility with gcc (over the builtin `clang` for macos) so doing this might solve #243

**Comment by molleweide** [comment]
> Get the gcc bin directory for
> * `brew info gcc`

hey does this really give you the bin dir for you? it doesn't for me but it whatever..

**Comment by molleweide** [comment]
> I've done some investigating into the issue, basically the treesitter parser doesn't compile with clang. The workaround referenced [here](https://github.com/nvim-neorg/neorg/issues/74#issuecomment-906627223) is to install it with `gcc` instead.
> 
> 1. Install gcc
>    
>    * `brew install gcc`
> 2. Get the gcc bin directory for
>    
>    * `brew info gcc`
> 3. Backup default gcc (it may not exist but worth doing just incase
>    
>    * `mv /usr/local/bin/gcc /usr/local/bin/gcc_old`
> 4. Make a symlink to `/usr/local/bin`
>    
>    * `ln -s /usr/local/Cellar/gcc/{{version_number}}/bin/gcc-11 /usr/local/bin/gcc`
> 5. Restart neovim with `neorg` enabled in doom_modules, let it install neorg parser for you
> 6. Remove gcc
>    
>    * `rm /usr/local/bin/gcc && mv /usr/local/bin/gcc_old /usr/local/bin/gcc`
>    * `brew uninstall gcc`
> 
> Let me know if this works for you! :)

I believe this fixed it for me!! Thank you very very very much.

**Comment by molleweide** [comment]
made it easier to copy and put in scripts

```bash
#!/usr/bin/env bash

# switch default gcc to brew installed one.

brew install gcc
mv /usr/local/bin/gcc /usr/local/bin/gcc_old
ln -s /usr/local/Cellar/gcc/11.2.0_1/bin/gcc-11 /usr/local/bin/gcc
```

```bash
#!/usr/bin/env bash

# revert back to system gcc

rm /usr/local/bin/gcc && mv /usr/local/bin/gcc_old /usr/local/bin/gcc
brew uninstall gcc
```

**Comment by connorgmeehan** [comment]
Warning this script will only work with gcc-11.2.0_1.  For future users check the version with 'brew info gcc' and modify the version number to match your version. :) can this be marked as resolved/closed @yesudeep?

**Comment by yesudeep** [comment]
Sure!

**Comment by yesudeep** [comment]
I wonder whether instead of creating symlinks, this could use the CC environment variable. (No bookkeeping required that way).

**Comment by connorgmeehan** [comment]
I've tried this one, it doesn't work because doom-nvim specifically sets the compiler to gcc (which resolves to clang 🤦‍♂️) and I think that overrides the ENV variable. We should revert this fix for v3.2.  Maybe it's worth detecting if clang is the compiler in path and notifying user that gcc is preferred, directing them to this issue?

**Comment by NTBBloodbath** [comment]
> I've tried this one, it doesn't work because doom-nvim specifically sets the compiler to gcc (which resolves to clang :man_facepalming:) and I think that overrides the ENV variable.

Hmm the main reason of why doom-nvim sets gcc as the compiler was due to some issues when compiling Haskell parser in MacOS too iirc. There's a PR in nvim-treesitter repo to fix the issues with Haskell and Neorg parsers but it's pretty much "abandoned" right now so maybe there's something missing in the PR changes.

> We should revert this fix for v3.2. Maybe it's worth detecting if clang is the compiler in path and notifying user that gcc is preferred, directing them to this issue?

We could try this, yeah. Sounds good to me :p
